### PR TITLE
[READY] Adds click shortcuts to xenobio console.

### PR DESCRIPTION
**Pull request by kriskog**
[Changelogs]: # (Your PR should contain a detailed changelog of notable changes, titled and categorized appropriately. This includes, new features, sprites, sounds, balance changes, admin tools, map edits, removals, big refactors, config changes, hosting changes and important fixes. An example changelog has been provided below for you to edit. If you need additional help, read https://github.com/tgstation/tgstation/wiki/Changelogs)

:cl: Skoglol
add: Added click shortcuts to xenobiology console. You no longer have to aim with the camera eye to do slime science.
/:cl:

[why]: # (Please add a short description [two lines down] of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.)

https://www.youtube.com/watch?v=zP5GP3sc0e0

Xenobiology is kind of a pain in its current state, where you gotta aim with the camera eye. The eye has inconsistent speed, and is prone to lag. This addition allows you to click the things you wish to manipulate instead, which is a lot smoother.

I've added a help action that explains the controls.
Threw together a icon for the button:

For now, shortcuts are:
Slime: Shift click slime to pick up, turf to drop all. Ctrl click to scan, alt click to feed potion.
Monkey: Ctrl click critted monkey to pick up, turf to drop a new one.

To do:
* [x] Rewrite it to use components instead. (Err, uses signals now. Good enough?)
* [x] Add some help text.

### CODE REVIEW

**Review by ShizCalev — CHANGES_REQUESTED** [review]
This should really be done via component instead of having a bunch of checks whenever ANYONE clicks on a turf/monkey/slime tbh.

**Comment by kriskog** [comment]
I'll look into components. 
It should go to parent and stop on the first if statement if you are not remote-controlling something though, which outside of using a camera console is rare.

**Comment by fludd12** [comment]
`back in my day, there was no console, and we got cores by bonesawing them out of slimes, As God Intended`

:+1:

**Comment by Cruix** [comment]
Evolution of xenobio:
> Slimes have to be fed and killed manually, and slime cores must be removed by surgery
> The slime recycler is added, removing the need for surgery
> The slime recycler is made to suck slimes in automatically so you don't have to load them manually
> The slime console is added
> The slime kill room is added
> The slime console now picks up things by clicking instead of moving

And that is not even to mention the new slime cores that were added over time. It is absolutely insane how much love this one niche, game-breakingly powerful sub-department gets.

**Comment by kriskog** [comment]
@ShizCalev I looked at components, and I see what you mean. I will be reworking this soon as I get the time.

**Comment by anconfuzedrock** [comment]
Xenobio is the lavaland of science

**Comment by Cyprex** [comment]
I think this PR is great, making the game less clunky is always nice. +1 From me.
That said, I do think xenobio nerfs are in order as well, something simple like making slimes take longer to grow or having them split into three instead of four might even suffice. These should be in a seperate PR however.

**Comment by epochayur** [comment]
@Cyprex good fucking luck doing that without getting mauled by whiny, entitled xeno mains, if #41276 is any indication

**Comment by subject217** [comment]
one man's ease of use is another man's powercreep. as others have mentioned, all you have to do is look at how xenobio has developed over time to see this in action.

File: code/modules/research/xenobiology/xenobio_camera.dm [hunk 1]
```diff
@@ -267,3 +267,124 @@
 			break
 	else
 		to_chat(owner, "<span class='notice'>Target is not near a camera. Cannot proceed.</span>")
+
```
Comment (MrDoomBringer) [inline code/modules/research/xenobiology/xenobio_camera.dm]: Could we add a tooltip or a help action button explaining these commands? Blue italics is what we use for OOC tips in descriptions and stuff iirc

**Review by MrDoomBringer — CHANGES_REQUESTED** [review]

File: code/modules/research/xenobiology/xenobio_camera.dm (hunk 1 above)
Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: That is on the to do list in the main post. These to_chats are copied from the relevant actions, if that's wrong (I'll take a look) then those should be changed as well.

**Comment by kriskog** [comment]
> 
> 
> one man's ease of use is another man's powercreep. as others have mentioned, all you have to do is look at how xenobio has developed over time to see this in action.

The current aieye object controls are broken, the speed is inconsistent and it has a tendency to lag. It is also quite clunky, to a point where even sorting slimes is a pain in the ass. If you want xenobio to be weaker, make a rebalance PR for it instead of arguing against improving the controls. "Balance" through clunky controls is a shit concept, and that is not something you can change my mind about.

**Comment by Cyprex** [comment]
I agree that clunky interfaces should not be a part of balance, consider AI multicamming where the same arguments were brought forth. Again, I'd like to see 
@epochayur I fail to see how the PR in question is an overall nerf, as linking the recycler to the console straight up increases the amount of monkeys you get when using the console(0.2->0.8 at t4). Later on you wouldn't even make new monkeys since you get 0.8 monkeys back per monkey, trivially. While dragging monkeys into an upgraded recycler isn't as good anymore, there is no need to since recycling is so effective. To add onto this I don't really see t4 recyclers used much anyway, since at that point you usually have plenty of gray slime cores, or even a reproductive/industrial gray extract.

**Comment by fludd12** [comment]
Please feel free to nerf crossbreeding at will - I've had balance worries since the start of the project.

Please don't make them useless, though. They do take a fair amount of effort to make, both in-game and code-wise ; ~ ;

**Comment by improvedname** [comment]
Now all we need is a wire that actives the slime proccesor, and conveyor belts autoloading into the smartfridge and gates that only allow young slimes to pass for true auto farming

**Comment by improvedname** [comment]
@cyprex the point of t4 monkey recycler was that you could turn 1 monkey into 256 in as minute

**Comment by Cyprex** [comment]
You need to kill the monkeys to recycle them, dragging a monkey out of the cells takes longer than dispensing a dozen grey slime extracts and injecting them with blood from a monkey corpse. Killing monkeys to recycle them would be an option, but I've never actually seen a setup like that, as the need for monkeys is just not there.

**Comment by kriskog** [comment]
@Cyprex With the change to camera consoles using upgraded parts for monkey recycling in that PR, you could just dump monkeys in the kill room and pick them up to duplicate them. That is the main reason I did the nerf in the same PR, as it clearly does not work.

File: code/modules/research/xenobiology/xenobio_camera.dm (hunk 1 above)
Comment (kriskog) [reply code/modules/research/xenobiology/xenobio_camera.dm]: Help action now added

**Comment by kriskog** [comment]
Okay, so I've made it use signals instead of all the checks. I needed the overloaded click procs to send new signals to get access to the objects clicked. They are still in the console file, what is the verdict on keeping them together and moving them to the owner files?

Currently the signals are registered and unregistered when the user opens or closes the console. 
I considered adding a component to the user when it opened the console, and removing it when they close it, but I couldn't quite get the removal to work. If that is the preferable solution, I would appreciate a quick outline of how that can be done.

**Comment by improvedname** [comment]
@cyprex you bs suprised how quick a electrified grille and a conveyor belt does the job

**Comment by kriskog** [comment]
I'll need a review on this before anything else is done, setting it ready.

**Comment by optimumtact** [comment]
kinda minor, but the signals in the future should ideally be generically named, rather than specific to the XENO, since there may be other things that want to listen to those clicks

**Pull request merged**